clash-cores/src/Clash/Cores/Internal/UART.hs

```diff
@@ -0,0 +1,317 @@
+{-|
+  Copyright   :  (C) 2021, LUMI GUIDE FIETSDETECTIE B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  Rowan Goemans <goemansrowan@gmail.com>
+
+  UART transmitter and receiver module
+-}
+
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE NumericUnderscores #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ConstraintKinds #-}
+
+module Clash.Cores.Internal.UART where
+
+import Clash.Prelude
+import Clash.Class.HasDomain (TryDomain, TryDomainResult(Found))
+import Data.Maybe (isJust)
+
+-- | Division that rounds the result
+type DivRound (a :: Nat) (b :: Nat) = Div (a + (Div b 2)) b
+
+-- | Utility type that represents a second
+type Second = 1_000_000_000_000 -- picoseconds
+
+-- | Calculate frequency from period at type level
+type PeriodToHz (period :: Nat) = DivRound Second period
```

Comment:
As noted [in this mailing list message](https://groups.google.com/g/clash-language/c/DYKLt_v_Xjg/m/PYgXRl31AQAJ), we have defined the term-level `hzToPeriod` to give the same result as `Div`, not `DivRound`. This was chosen to keep the type-level thing simple, arguing that if you need the picosecond precision that is the difference between them, you should be specifying a period rather than a rate. This is also in the 1.4.0 [changelog](https://hackage.haskell.org/package/clash-prelude-1.4.0/changelog), if you search for 1253, you'll find:
> * `hzToPeriod` now takes a `Ratio Natural` rather than a `Double`. It rounds slightly differently, leading to more intuitive results and satisfying the requested change in #1253. Clash expresses clock rate as the clock period in picoseconds. If picosecond precision is required for your design, please use the exact method of specifying a clock period rather than a clock frequency.

So if we add `HzToPeriod` to our Prelude, it will most likely be just `Div`, not `DivRound`. `PeriodToHz` is actually the same function, though they obviously do not round-trip because of rounding.

I suggest you use just `Div` here. I don't expect `UartBaudGenTick` to create a worse tick for it, right?